framework/src/controller/application.js

```diff
@@ -233,10 +252,31 @@ class Application {
 		);
 
 		assert(
-			!Object.keys(this.getTransactions()).includes(Transaction.TYPE),
+			!Object.keys(this.getTransactions()).includes(
+				Transaction.TYPE.toString()
+			),
 			`A transaction type "${Transaction.TYPE}" is already registered.`
 		);
 
+		const Base = getBasePrototype(Transaction);
+
+		assert(
+			Base.name === BaseTransaction.name,
+			'Transaction must extend BaseTransaction.'
+		);
+
+		const publicPropertiesOfBase = getPublicProperties(Base);
+		const publicPropertiesOfBaseTransaction = getPublicProperties(
+			BaseTransaction
+		);
+
+		assert(
+			publicPropertiesOfBaseTransaction.every(prop =>
+				publicPropertiesOfBase.includes(prop)
+			),
+			'Transaction must use compatible BaseTransaction.'
+		);
```

Comment:
Suggest to use the same pattern we used to validate other object. `JSON-Schema`.

Define the full JSON-Schema for what you expect a transaction to be look like and then match with it. 

This will help to check the types as well:

https://github.com/epoberezkin/ajv-keywords#types

This package is already in the dependency tree, so safe to use.

Reply:
```js
{
	instanceOf: 'BaseTransaction',
	required: ['apply', 'undo'],
	properties: {
		apply: {
			typeof: 'function',
		},
		undo: {
			typeof: 'function',
		},
	},
};
```

Reply:
`instanceOf` will not work unless we create custom one with instance, but it defeats the purpose of not using exact match.

other than that, I think we can define all the function/properties with types, but I think this is too redundant in a way that all of the type are defined in the TypeScript already.
also, for property of instance, it has it's own validation

therefore, I think checking dynamically make more sense.

Reply:
I am still not convinced on doing it this way. Using `Object.getOwnPropertyNames` to match a class object is seems weird and not completely under the hood. Unless we have descriptive details of what we are matching, things will get difficult to understand later. And would also be difficult to refer to user that what a transaction could be, and what are required interfaces to implement and what are optional. If you have a JSON-Schema that will make it very clear and descriptive.

Also the this code  `Object.getPrototypeOf(instance);` and `Base.name === BaseTransaction.name`  is deceptive. It refers to the fact Transaction can be extended by any thing which name is `BaseTransaction`. So if we have the code below then it will pass that check, although its invalid BaseTransaction. 

```
class BaseTransaction(){}
class MyTransaction extends BaseTransaction{}
```

> That all of the type are defined in the TypeScript already.

The above reasoning is not correct as well under the context. User is not bound to use typescript where transpiler check and validate the types. Even if he does type matching, then question will come back to root point, against which BaseTransaction it will match, one which is part of the Framework or one which is in the transactions package. 

I suggest this issue need more thought process to define a correct pattern to handle breaking changes in transaction.

Reply:
I would tend to highlight my initial thought here as well. We just need a check that transaction is extended by `BaseTransaction` which `lisk-sdk` supports and exports. All custom transactions should must be extended by that one only. 

If someone have any client side logic or standalone script using `lisk-transactions` only, that's not an issue. But for any custom application and transaction for the application one must and only user the `lisk-sdk` package. 

@MaciejBaj

Reply:
true, we shouldn't need to check if the class extends base transaction or not.
At this point it just requires to have exact same public signature.
1. we cannot check static type in the schema. we need another schema, but it's checked already so it should be ok.
2. we should not check the property, because we don't know when it will be assigned from the class. Also, possible that it requires some parameter in the constructor for some specific custom transaction.

Let's try with schema even though it is redundant.
If it causes too much trouble, I will move it back to dynamic check later time.